server/requirements.txt

```diff
@@ -18,4 +18,9 @@ aiomcache==0.6.0
 uvloop==0.14.0
 bootstrapped==0.0.2
 tqdm==4.46.1
-slackclient[optional]==2.7.1
\ No newline at end of file
+slackclient[optional]==2.7.1
+gcloud-aio-kms==3.2.0
+# gcloud-aio-kms does not pin the exact versions of the following packages
```

Comment:
`requirements.txt` misusage 101... https://github.com/talkiq/gcloud-aio/blob/master/auth/requirements.txt
I am also fighting to fix the installation https://github.com/talkiq/gcloud-aio/pull/160

Reply:
FYI: as a general rule, libraries should pin compatible version _ranges_ and apps should pin _specific_ versions: you can read a lot more about reasoning why [eg. here](https://stackoverflow.com/questions/28509481/should-i-pin-my-python-dependencies-versions) but the main advantage is that it helps app developers solve conflicts -- for example, if `slackclient[optional]` had depended on a specific older version of `backoff`, everything still would have worked out since we've already verified that all of the 1.x releases of `backoff` are compatible with `gcloud-aio`. :)

Reply:
All is correct except that you seem to confuse `setup.py`'s `install_requires` and `requirements.txt` @TheKevJames

https://packaging.python.org/discussions/install-requires-vs-requirements/

Reply:
The difference is the same as between `go.mod` and `go.lock` (to lesser extent), `package.json` and `yarn.lock`, `Cargo.toml` and `Cargo.lock`, `Gemfile` and `Gemfile.lock`, etc.

Reply:
Huh, I'm confused -- when you install `gcloud-aio-kms` here you're using the `install_requires`, which according to your instructions is meant to be version ranges, as we have it. So... from the perspective of `athenian-api` or any downstream app, you should _never_ get pinned versions from your libraries.

If I'm reading that link correctly, the only folks who should ever use the `requirements.txt` file are folks developing `gcloud-aio` itself. In which case... doesn't it still do what it should, since it let's us verify that the latest compatible versions are guaranteed compatible?

We always pin all dependencies for applications, but pinning them for libraries doesn't make sense to me -- if we had them pinned for our local developement, wouldn't that just hide potential version conflicts with newer releases and make it harder for us to detect issues before users do?

Reply:
Look: this is what you do in `setup.py` - this is a well known Python packaging gotcha: https://github.com/talkiq/gcloud-aio/blob/master/auth/setup.py#L24 What you should do instead is to copy your current `requirements.txt` to `setup.py` and write a new `requirements.txt` with all the versions pinned. When you `pip install`, it reads `setup.py`'s `install_requires` by default. BTW, I was the same confused fellow 7 years ago, I did it [here](https://github.com/Samsung/veles/blob/master/requirements.txt), for example. Learn from my mistakes :D

The reason why I don't pin all the nested deps in my project's `requirements.txt` is just because I am too lazy and not always `pip freeze`. BTW, `pip freeze` is exactly for this: to pin everything to your `requirements.txt`. The reason for pinning is to have reproducible builds. There are numerous cases when people are not attentive enough to semantic versioning and release breaking changes in minor or patch version bumps. The more deps you have, the more likely the chance is you'll get a broken build. Heck, even scipy has recently broken backward compatibility in 1.2, those guys took a big shit on semantic versioning. Besides, you'll not be able to test with all possible version combinations if your project is big enough. You'll get a combinatorial explosion of all possible version sets. 

> if we had them pinned for our local developement, wouldn't that just hide potential version conflicts with newer releases

This is why we have @dependabot and @snyk who automatically submit version bumping and security fixing PRs to your `requirements.txt` and `setup.py`.